src/persist/src/indexed/mod.rs

```diff
@@ -73,6 +73,81 @@ impl PendingResponse {
     }
 }
 
+struct Pending {
+    // Only drained by drain_pending_writes.
+    pending_writes: HashMap<Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>>,
+    // Only drained by drain_pending.
+    pending_responses: Vec<PendingResponse>,
+    pending_seals: HashMap<Id, u64>,
+}
+
+impl Pending {
+    fn new() -> Self {
+        Self {
+            pending_writes: HashMap::new(),
+            pending_responses: Vec::new(),
+            pending_seals: HashMap::new(),
+        }
+    }
+
+    fn validate_empty(&self) -> Result<(), Error> {
+        if !self.pending_writes.is_empty() {
+            return Err(Error::from(format!(
+                "still have {} pending writes after draining pending writes, expected 0.",
+                self.pending_writes.len()
+            )));
+        }
+
+        if !self.pending_responses.is_empty() {
+            return Err(Error::from(format!(
+                "still have {} pending responses after draining pending responses, expected 0.",
+                self.pending_responses.len()
+            )));
+        }
+
+        if !self.pending_seals.is_empty() {
+            return Err(Error::from(format!(
+                "still have {} pending seals after draining pending seals, expected 0.",
+                self.pending_seals.len()
+            )));
+        }
+
+        Ok(())
+    }
+
+    fn add_writes(&mut self, updates: Vec<(Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>)>) {
+        for (id, updates) in updates {
+            self.pending_writes.entry(id).or_default().extend(updates);
+        }
+    }
+
+    fn add_response(&mut self, resp: PendingResponse) {
+        self.pending_responses.push(resp);
+    }
+
+    fn add_seals(&mut self, ids: Vec<Id>, seal: u64) {
+        for id in ids {
+            self.pending_seals.insert(id, seal);
+        }
+    }
+
+    fn writes(&mut self) -> HashMap<Id, Vec<((Vec<u8>, Vec<u8>), u64, isize)>> {
+        std::mem::take(&mut self.pending_writes)
+    }
+
+    fn responses(&mut self) -> Vec<PendingResponse> {
+        std::mem::take(&mut self.pending_responses)
+    }
+
+    fn seals(&mut self) -> HashMap<Id, u64> {
+        std::mem::take(&mut self.pending_seals)
+    }
+
+    fn has_responses(&self) -> bool {
```

Comment:
I'm not convinced that these accessors that get by taking and this has_responses API is the best idea ever

Reply:
Agree but it seems good enough. If you're worried about it, leave a TODO so we know it's something to keep thinking about